jetty-server/src/main/java/org/eclipse/jetty/server/DeflaterPool.java

```diff
@@ -0,0 +1,112 @@
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.server;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.zip.Deflater;
+
+public class DeflaterPool
+{
+    private final Queue<Deflater> _pool;
+    private final int _compressionLevel;
+    private final boolean _nowrap;
+    private int _capacity;
+    private AtomicInteger _numDeflaters = new AtomicInteger(0);
+
+    /**
+     * Create a {@link Deflater} Pool
+     *
+     * If given a capacity less or equal to zero the Deflaters will not be pooled
+     * and will be created on acquire and ended on release.
+     *
+     * @param capacity maximum number of Deflaters which can be contained in the pool
+     * @param compressionLevel the default compression level for new Deflater objects
+     * @param nowrap if true then use GZIP compatible compression for all new Deflater objects
+     */
+    public DeflaterPool(int capacity, int compressionLevel, boolean nowrap)
+    {
+        _capacity = capacity;
+        _compressionLevel = compressionLevel;
+        _nowrap = nowrap;
+
+        if (_capacity > 0)
+            _pool = new ConcurrentLinkedQueue<>();
+        else
+            _pool = null;
+    }
+
+    protected Deflater newDeflater()
+    {
+        return new Deflater(_compressionLevel, _nowrap);
+    }
+
+    /**
+     * @return Deflater taken from the pool if it is not empty or a newly created Deflater
+     */
+    public Deflater acquire()
+    {
+        if (_pool == null)
+            return newDeflater();
+
+        Deflater deflater = _pool.poll();
+        if (deflater == null)
+            deflater = newDeflater();
+        else
+        {
+            _numDeflaters.decrementAndGet();
+        }
+
+        return deflater;
+    }
+
+    /**
+     * @param deflater returns this Deflater to the pool or calls deflater.end() if the pool is full.
+     */
+    public void release(Deflater deflater)
+    {
+        if (deflater == null)
+            return;
+
+        if (_pool == null)
+        {
+            deflater.end();
+            return;
+        }
+
+        while(true)
+        {
+            int d = _numDeflaters.get();
+
+            if (d >= _capacity)
+            {
+                deflater.end();
+                break;
+            }
+
+            if (_numDeflaters.compareAndSet(d,d+1))
+            {
+                deflater.reset();
+                _pool.add(deflater);
+                break;
+            }
+        }
```

Comment:
This `while` loop is not optimal. A large number of threads may enter it, find a larger than capacity `d` and empty the pool.

Seems overcomplicated to me.
Ins't this simpler:

```java
deflater.reset();
if (!_pool.offer(deflater))
    deflater.end();
```

Reply:
Isn't there a need to empty the internal buffers before calling `.end()`?